decisions/websre/0004-use-1-kubernetes-cluster-per-environment.md

```diff
@@ -0,0 +1,56 @@
+
+At Mozilla, our service environments generally include Development (Dev), Staging (Stage), and Production (Prod). There can sometimes also be a Testing environment (usually ephemeral). Generally, these environments are in the same infrastructure provider account, and differ at the infrastructure or application platform level (e.g. at the Kubernetes level) - with the exception of Testing, which tends to be ephemeral within a particular context (e.g. a docker-compose spin up or a kind cluster in a CI pipeline).
+
+There are multiple ways that we differentiate these persistent (dev, stage, prod) environments on our infrastructure and application platforms. Within a Kubernetes context, some common patterns for separating persistent environments are:
+1. Having a distinct Kubernetes cluster per environment;
+2. Having one namespace per environment in a shared Kubernetes cluster;
+3. Having distinct node pool(s) per environment within a single cluster along with using selectors, taints, and tolerations.
+
+## Decision
+
+For Web SRE Services persistent environments - dev, stage, prod - that are Kubernetes-based, we will:
+* Use a distinct Kubernetes cluster per environment (dev, stage, prod) within the same infrastructure account;
+* Use a distinct Kubernetes namespace within each cluster per application;
+* Deploy applications to these distinct clusters as relevant to the expected application environment (e.g. applications should deploy to the stage cluster, prod applications should deploy to the prod cluster, etc.);
+* Deploy operationals & infrastructure changes & upgrades to these distinct clusters, specifically moving changes from stage to prod within an appropriate timeframe, with an end goal of keeping stage and prod clusters in close alignment;
+* Development clusters availability and uptime expectations (of the application or the infrastructure) does not include support outside of business hours;
+* Development clusters can optionally include automatic infrastructure or platform updates, as a first place to review these changes and their effects;
+* Only use reviewed & approved, production-ready stable application & infrastructure releases on the Production cluster;
+* Only use reviewed & approved, stable infrastructure releases on the Stage cluster, and deploy those infrastructure changes when able to relatively quickly move the changes from stage into prod to avoid divergent results from staging and production applications.
+
+This ADR does not include the following:
+* How the applications within each cluster name their namespaces (e.g. keeping "discourse_prod" namespace in a production cluster is redundent, but also very explicit, and is left up to the SREs working on the service);
+* An immediate push to migrate all existing services to this pattern, as it will cause a lot of churn. Instead, this should be used going forward when doing migrations or starting new services;
```

Comment:
Do we have any existing services whose environment is _not_ a subset of { dev, stage, prod, ephemeral_test }?

Reply:
possibly.